crates/matrix-sdk/src/room/mod.rs

```diff
@@ -1313,11 +1319,11 @@ impl Room {
     ///
     /// This method makes sure the room that was returned when joining a room
     /// has been echoed back in the sync.
-    /// Warning: This waits until a sync happens and does not return if no sync
+    /// Warning: This blocks until a sync happens and does not return if no sync
     /// is happening! It can also return early when the room is not a joined
     /// room anymore!
     #[instrument(skip_all)]
-    pub async fn sync_up(&self) {
+    pub fn sync_up(&self) {
         while !self.is_synced() && self.state() == RoomState::Joined {
             self.client.inner.sync_beat.listen().wait_timeout(Duration::from_secs(1));
```

Comment:
So we are waiting in a blocking manner here? That seems problematic.

Reply:
Yep! In our code it's only used in testing, so we should probably make it test-only because it seems dangerous for public use. Also I'm wondering if it's not broken, because it never returns in some test...

Reply:
The tokio test timeline uses a single-threaded runtime. That means if a task runs this code in a test, it will definitely always time out since there is no other free runtime thread to poll any other task that could unblock this one.

Reply:
The test contains a `worker_threads=4` annotation, so we should be fine with respect to the number of threads?

Reply:
I think it should be doable with a `tokio::select!`: either timeout using `tokio::time::sleep` or await for the sync_beat to notify.